packages/roosterjs-editor-api/lib/paste/buildClipBoardData.ts

```diff
@@ -0,0 +1,149 @@
+import { Editor } from 'roosterjs-editor-core';
+import { convertInlineCss, fromHtml } from 'roosterjs-editor-dom';
+import getFormatState from '../format/getFormatState';
+import processImages from './utils/processImages';
+
+const INLINE_POSITION_STYLE = /(<\w+[^>]*style=['"][^>]*)position:[^>;'"]*/gi;
+const TEXT_WITH_BR_ONLY = /^[^<]*(<br>[^<]*)+$/i;
+const COMMENT = /<!--([\s\S]*?)-->/gi;
+const HTML_REGEX = /<html[^>]*>[\s\S]*<\/html>/i;
+
+interface WindowForIE extends Window {
+    clipboardData: DataTransfer;
+}
+
+/**
+ * Build ClipboardData from a paste event
+ * @param event The paste event
+ * @param editor The editor
+ * @param callback Callback function when data is ready
+ * @param useDirectPaste Whether use direct HTML instead of using temp DIV
+ */
+export default function buildClipBoardData(
+    event: ClipboardEvent,
+    editor: Editor,
+    callback: (clipboardData: ClipBoardData) => void,
+    unsafeHtmlFilter?: (html: string) => string,
+) {
+    let dataTransfer =
+        event.clipboardData || (<WindowForIE>editor.getDocument().defaultView).clipboardData;
+    let clipboardData: ClipBoardData = {
+        snapshotBeforePaste: null,
+        currentFormat: getFormatState(editor),
+        imageData: {
+            file: getImage(dataTransfer),
+        },
+        textData: dataTransfer.getData('text'),
+        htmlData: null,
+        htmlFragment: null,
+    };
+    let retrieveHtmlCallback = createCallback(editor, clipboardData, unsafeHtmlFilter, callback);
+
+    if (!unsafeHtmlFilter || !directRetrieveHtml(event, editor, retrieveHtmlCallback)) {
+        retrieveHtmlViaTempDiv(editor, retrieveHtmlCallback);
+    }
+}
+
+function createCallback(
+    editor: Editor,
+    clipboardData: ClipBoardData,
+    unsafeHtmlFilter: (html: string) => string,
+    pasteCallback: (data: ClipBoardData) => void) {
+    return (html: string) => {
+        let matches = HTML_REGEX.exec(html);
+        html = matches ? matches[0] : html;
+
+        if (unsafeHtmlFilter) {
+            html = unsafeHtmlFilter(html);
+        }
+        html = convertInlineCss(html);
+        html = normalizeContent(html);
+
+        // 4. Prepare HTML fragment for pasting
+        let document = editor.getDocument();
+        let fragment = document.createDocumentFragment();
+        let nodes = fromHtml(html, document);
+        for (let node of nodes) {
+            fragment.appendChild(node);
+        }
+
+        clipboardData.htmlData = html;
+        clipboardData.htmlFragment = fragment;
+
+        processImages(clipboardData);
+        pasteCallback(clipboardData);
+    }
+}
+
+function getImage(dataTransfer: DataTransfer): File {
+    // Chrome, Firefox, Edge support dataTransfer.items
+    let fileCount = dataTransfer.items ? dataTransfer.items.length : 0;
+    for (let i = 0; i < fileCount; i++) {
+        let item = dataTransfer.items[i];
+        if (item.type && item.type.indexOf('image/') == 0) {
+            return item.getAsFile();
```

Comment:
What if there are multiple images? Or do they come together as one file?

Reply:
It won't.

For example, if copy from OneNote when select two images, OneNote will generate a screenshot which contains both images and set it into clipboard.

In some other cases, say copy from browser with 2 images, there will be no image set into clipboard, but only some html text.

Reply:
Got it.